### ISSUE TITLE
GPU support for BlockNonlinearForm (memory sync in BlockNonlinearForm::Mult)

### ISSUE THREAD

**Issue by jhbaek04** [mfem/mfem#4102]
Hello,

While I try a multi-variable problem using `BlockNonlinearForm` on CUDA, I've observed that including `by.SyncFromBlocks();` at the end of `BlockNonlinearForm::Mult()` is necessary. Without this additional call, the `NewtonSolver` fails to converge.  I think this line ensures memory sync after processing the residual entries associated with the Dirichlet DOFs.

Could you confirm if my understanding is correct?

Thank you,
Jon

**Comment by v-dobrev** [comment]
Yes, I think you are right -- in general, at the end of `BlockNonlinearForm::Mult()`, one needs to ensure that the validity flags of `by` (monolithic) is sync'd from its sub-vectors. In addition, in general, one has to make sure that the validity flags of `by` (monolithic) are propagated to `y` (since `by` is an alias for `y`). There's also the possibility that `pby` may need to be sync'd with `ys`.

So, yes, it looks like we need to revisit the synchronizations in `BlockNonlinearForm::Mult()` (and the parallel version as well, probably) to ensure the validity flags are propagated correctly.

Looking at the code, I also wonder if we really need the `xs` and `ys` `BlockVector`s -- it looks like we can simply use `pbx` and `pby`. @jandrej, does that seem correct? If we don't use `xs` and `ys` that will also simplify the necessary synchronizations.

**Comment by jhbaek04** [comment]
Thank you for the clarification and the details. For my case, it seems one synchronization of `by` at the end of `BlockNonlinearForm::Mult()` ensures a proper sync of `y`, but I am not sure if this is always true.

I've additionally tested a case with an identity prolongation matrix to more thoroughly check `BlockNonlinearForm::Mult()`, and it seems that `aux1.SyncFromBlocks();` is needed after applying the prolongation to `bx` in `BlockNonlinearForm::Prolongate()`. I might be wrong, but I think we are missing synchronization there as well.

**Comment by v-dobrev** [comment]
Since `aux1` is internal, I'm not sure we really need to ensure its monolithic Vector is in sync with its sub-vectors. For example, if we only use its sub-vectors to perform operations, then sync with the monolithic vector is probably not necessary. Do we actually use the monolithic version of `aux1` anywhere? 

On the other hand, for the output parameter `y`, the sync is clearly needed since the caller does not see the sub-blocks we create to perform operations.
